### PR TITLE
8279258: Auto-vectorization enhancement for two-dimensional array operations

### PR DESCRIPTION
**Pull request by DamonFool**
Hi all,

Happy Christmas Day!

We have observed that C2 fails to auto-vectorize two-dimensional array operations in our machine learning programs.
And we have made an reproducer in the JBS.

Now let's discuss the reproducer.
The auto-vectorization fails due to `cl->slp_max_unroll() == 0` [1], which means the previous slp analysis never passed.

As for our example, C2 had tried its first slp analysis with `future_unroll_cnt=4` [2].
But unfortunately, it failed due to the loop IR is too complicated [3] like the following.
```
SuperWord::transform_loop: loop too complicated, cl_exit->in(0) != lpt->_head
cl_exit 823 823  CountedLoopEnd  ===  738  822  [[ 907  682 ]] [lt] P=0.999999, C=-1.000000 !orig=[680]
cl_exit->in(0) 738 738  IfTrue  ===  735  [[ 823 ]] #1 !orig=[442] !jvms: DoubleArray2::test @ bci:17 (line 10)
lpt->_head 1267 1267  CountedLoop  ===  1267  1224  682  [[ 1267  1278  1283  1284  1288  1254  1282 ]] inner stride: 2 main of N1267 !orig=[824],[748],[687] !jvms: DoubleArray2::test @ bci:30 (line 11)
    Loop: N1267/N682  counted [int,int),+2 (65 iters)  main rc  has_sfpt rce
RangeCheck       Loop: N1267/N682  counted [int,int),+2 (65 iters)  main rc  has_sfpt rce
Unroll 4         Loop: N1267/N682  counted [int,int),+2 (65 iters)  main rc  has_sfpt rce
Loop: N0/N0  has_sfpt
  Loop: N493/N463  limit_check profile_predicated predicated counted [0,int),+1 (65 iters)  sfpts={ 453 }
    Loop: N946/N966  counted [0,int),+1 (4 iters)  pre has_sfpt
    Loop: N1483/N682  counted [int,int),+4 (65 iters)  main rc  has_sfpt
    Loop: N857/N877  counted [int,int),+1 (4 iters)  post has_sfpt
PredicatesOff
```

Then, C2 unrolled the loop with `unroll-factor=4` and also did some other opts, which actually simplified the loop IR representation.

And then, comes the next round of loop unrolling analysis, in which C2 would check if `future_unroll_cnt=8` [2] is OK for unrolling.
C2 rejected `future_unroll_cnt=8` for this example and returned false immediately [4] without doing a second slp analysis, leaving `cl->slp_max_unroll() == 0`.
But if we re-do the slp analysis with `future_unroll_cnt=4` before returning false, it would pass.

So the key idea is:
```
  slp analysis may fail due to the loop IR is too complicated especially during the early stage of loop unrolling analysis.
  But after several rounds of loop unrolling and other optimizations, it's possible that the loop IR becomes simple enough to pass the slp analysis.
  So C2 can try one more slp analysis instead of returning false immediately here [4].
```

We have observed up to 1.7x performance improvement by our micro benchmarks.

image

Testing:
  - tier1 ~ tier3 on Linux/x64, no regression.

Thanks.
Best regards,
Jie

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/superword.cpp#L129
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/loopTransform.cpp#L908
[3] https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/superword.cpp#L137
[4] https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/loopTransform.cpp#L910

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279258](https://bugs.openjdk.java.net/browse/JDK-8279258): Auto-vectorization enhancement for two-dimensional array operations

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to 1a9b4c847acf47ad9174c53bef89fc8db5756190
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6933/head:pull/6933` \
`$ git checkout pull/6933`

Update a local copy of the PR: \
`$ git checkout pull/6933` \
`$ git pull https://git.openjdk.java.net/jdk pull/6933/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6933`

View PR using the GUI difftool: \
`$ git pr show -t 6933`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6933.diff">https://git.openjdk.java.net/jdk/pull/6933.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -907,7 +909,18 @@ bool IdealLoopTree::policy_unroll(PhaseIdealLoop *phase) {
       future_unroll_cnt > LoopUnrollMin &&
       (future_unroll_cnt - 1) * (100.0 / LoopPercentProfileLimit) > cl->profile_trip_cnt() &&
       1.2 * cl->node_count_before_unroll() < (double)_body.size()) {
-    return false;
+    if (UseSuperWord && (cl->slp_max_unroll() == 0) &&
+        (cl->unrolled_count() - 1) * (100.0 / LoopPercentProfileLimit) <= cl->profile_trip_cnt()) {
```
Comment (neliasso) [inline src/hotspot/share/opto/loopTransform.cpp:913]: On line 911 and 913 this is repeated: 
"(X - 1) * (100.0 / LoopPercentProfileLimit) > cl->profile_trip_cnt()"
Please replace that with a method.

**Review by neliasso — COMMENTED** [review]
Hi Jie,

I hope you have a good holiday too!

Nice find, and straight forward fix too. I have one comment in the code.

It's excellent that you have added a microbenchmark too. I would like to have a small regression test too, that quickly fails if this would break in the future. Perhaps something using the IR Testing framework.

Best regards,
Nils Eliasson

**Comment by DamonFool** [comment]
> Hi Jie,
> 
> I hope you have a good holiday too!
> 
> Nice find, and straight forward fix too. I have one comment in the code.
> 
> It's excellent that you have added a microbenchmark too. I would like to have a small regression test too, that quickly fails if this would break in the future. Perhaps something using the IR Testing framework.
> 

Thanks @neliasso !

A regression test has been added and your comment has been addressed.
Thanks.

**Review by neliasso — COMMENTED** [review]
Excellent!

I will run some tests and if they pass I am ready to approve this.

Regards,
Nils

**Comment by neliasso** [comment]
Testing tier 1 to 3 passed.

**Comment by DamonFool** [comment]
> Testing tier 1 to 3 passed.

Thanks @neliasso for the review and testing.

So are you fine with this change?
Thanks.

**Review by neliasso — APPROVED** [review]
Yes. Looks good!

**Review by vnkozlov — COMMENTED** [review]
Please, wait after Christmas break.
I am planning to look on it. You do need second review.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 2]
```diff
@@ -899,15 +899,28 @@ bool IdealLoopTree::policy_unroll(PhaseIdealLoop *phase) {
...
   //   the residual iterations are more than 10% of the trip count
   //   and rounds of "unroll,optimize" are not making significant progress
   //   Progress defined as current size less than 20% larger than previous size.
   if (UseSuperWord && cl->node_count_before_unroll() > 0 &&
       future_unroll_cnt > LoopUnrollMin &&
-      (future_unroll_cnt - 1) * (100.0 / LoopPercentProfileLimit) > cl->profile_trip_cnt() &&
+      is_residual_iters_large(future_unroll_cnt, cl) &&
       1.2 * cl->node_count_before_unroll() < (double)_body.size()) {
-    return false;
+    if (UseSuperWord && (cl->slp_max_unroll() == 0) &&
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopTransform.cpp:912]: UseSuperWord was already checked above.

Comment (DamonFool) [reply src/hotspot/share/opto/loopTransform.cpp]: > UseSuperWord was already checked above.

Thanks @vnkozlov for your review.
The redundant `UseSuperWord` check had been removed.

**Comment by DamonFool** [comment]
gc/metaspace/TestMetaspacePerfCounters.java#id3 failed on Linux/x86_32.

But it does pass with the previous pre-submit tests and my local tests.
So I don't think it was caused by this change.

**Comment by DamonFool** [comment]
> Please, wait after Christmas break. I am planning to look on it. You do need second review.

Hi @vnkozlov , I've merged the master and updated the copyright year to 2022.
Thanks.

**Comment by vnkozlov** [comment]
May understanding was that we do SLP unrolling analysis only after we did all RCE (range check elimination) and other loop optimizations so that only unrolling is left. But this is not the case. As your output and my testing show we have RCE after we tried SLP analysis:
```
RangeCheck       Loop: N1267/N682  counted [int,int),+2 (65 iters)  main rc  has_sfpt rce
```
Simply checking `is_unroll_only()` does not help currently:
```
     // Only attempt slp analysis when user controls do not prohibit it
-    if (LoopMaxUnroll > _local_loop_unroll_factor) {
+    if (cl->is_unroll_only() && (LoopMaxUnroll > _local_loop_unroll_factor)) {
       // Once policy_slp_analysis succeeds, mark the loop with the
       // maximal unroll factor so that we minimize analysis passes
```

because `HasRangeChecks` flag is still set even after RCE:
```
    Loop: N1483/N682  counted [int,int),+4 (65 iters)  main rc  has_sfpt
```

Can you investigate why  `HasRangeChecks` flag is still set?

**Comment by vnkozlov** [comment]
We specifically have next code to see if RCE removed all range checks from loop:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/loopTransform.cpp#L2984

**Comment by DamonFool** [comment]
> Can you investigate why `HasRangeChecks` flag is still set?

This is because once HasRangeChecks flag is set, it will never be cleared.

Proposed fix:
```
diff --git a/src/hotspot/share/opto/loopTransform.cpp b/src/hotspot/share/opto/loopTransform.cpp
index beb6c5a2830..e0b2fb77c35 100644
--- a/src/hotspot/share/opto/loopTransform.cpp
+++ b/src/hotspot/share/opto/loopTransform.cpp
@@ -3540,6 +3540,8 @@ bool IdealLoopTree::iteration_split_impl(PhaseIdealLoop *phase, Node_List &old_n
     if (should_rce) {
       if (phase->do_range_check(this, old_new) != 0) {
         cl->mark_has_range_checks();
+      } else {
+        cl->clear_has_range_checks();
       }
     } else if (PostLoopMultiversioning) {
       phase->has_range_checks(this);
diff --git a/src/hotspot/share/opto/loopnode.hpp b/src/hotspot/share/opto/loopnode.hpp
index 4e6a56f60ba..18145a51839 100644
--- a/src/hotspot/share/opto/loopnode.hpp
+++ b/src/hotspot/share/opto/loopnode.hpp
@@ -114,6 +114,7 @@ public:
   void mark_loop_vectorized() { _loop_flags |= VectorizedLoop; }
   void mark_has_atomic_post_loop() { _loop_flags |= HasAtomicPostLoop; }
   void mark_has_range_checks() { _loop_flags |=  HasRangeChecks; }
+  void clear_has_range_checks() { _loop_flags &= ~HasRangeChecks; }
   void mark_is_multiversioned() { _loop_flags |= IsMultiversioned; }
   void mark_strip_mined() { _loop_flags |= StripMined; }
   void clear_strip_mined() { _loop_flags &= ~StripMined; }
```

But I would suggest fix it in a separate JBS issue.
What do you think?

**Comment by DamonFool** [comment]
> // Only attempt slp analysis when user controls do not prohibit it
> -    if (LoopMaxUnroll > _local_loop_unroll_factor) {
> +    if (cl->is_unroll_only() && (LoopMaxUnroll > _local_loop_unroll_factor)) {
>        // Once policy_slp_analysis succeeds, mark the loop with the
>        // maximal unroll factor so that we minimize analysis passes

Hi @vnkozlov ,

If I understand it correctly, we should not make this change.

This is because `mark_do_unroll_only()` is only performed in `SuperWord::output()`.
And we can only invoke `SuperWord::output()` if and only if SLP unrolling analysis has paased.
What do you think?

**Comment by vnkozlov** [comment]
Yes, we can't use `is_unroll_only()` but we can use `!cl->range_checks_present()` with your `clear_has_range_checks()` change. It bails early instead of bailing after check `(cl_exit->in(0) != lpt->_head)` in `superword.cpp` [3].

Doing all this  we would still need your original changes because we hit the same issue again - rejection for future_unroll_cnt=8 because of low number of iterations (64).

Your changes are conservative and well. I will run few experiments to see if number of created vectors is changed and how.

**Comment by vnkozlov** [comment]
I run `compiler/c2/cr6340864/ compiler/codegen/ compiler/loopopts/superword/ compiler/vectorization` tests with `-Xbatch -XX:-TieredCompilation -XX:CICompilerCount=1 -XX:+TraceNewVectors` flags and don't see any changes in number of created vectors. Good.
I will submit formal testing and run performance testing too.

**Comment by vnkozlov** [comment]
Changes I tested includes your original changes, `clear_has_range_checks()` and  `!cl->range_checks_present()`.

**Comment by DamonFool** [comment]
> Changes I tested includes your original changes, `clear_has_range_checks()` and `!cl->range_checks_present()`.

Thanks @vnkozlov for your review and testing.
Patch had been updated.

**Review by vnkozlov — APPROVED** [review]
Testing results looks good. But we have to wait performance results - could be tomorrow.

**Comment by vnkozlov** [comment]
Performance results are good - no regressions.
You can integrate.
Thanks!

**Review by vnkozlov — CHANGES_REQUESTED** [review]
I forgot. Before integration, please rename tests. `AutoVectorization.java` and `TestAutoVectorization.java` are too general.

**Comment by DamonFool** [comment]
> I forgot. Before integration, please rename tests. `AutoVectorization.java` and `TestAutoVectorization.java` are too general.

Hi @vnkozlov ,

I had renamed the tests by adding the suffix `2DArray`.
Are you fine with it?

Or do you have any suggestions?
Thanks.

**Review by vnkozlov — APPROVED** [review]
Good. You can push it now.

**Comment by DamonFool** [comment]
> Good. You can push it now.

Thanks @vnkozlov and @neliasso .

/integrate

**Pull request closed without merging**